File: leetcode-in-python/solution/102_binary_tree_level_order_traversal.py
from typing import List
from datastruct.treenode import TreeNode
import unittest


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        return list(self._levelOrder(root))

    def _levelOrder(self, root: TreeNode):
        if root is None:
            return
        node = root
        queue = [node]
        size = len(queue)
        while size > 0:
            level_nodes = []
            # 当前层节点大小
            size = len(queue)
            for i in range(size):
                node = queue.pop(0)
                level_nodes.append(node.val)
                # 左子树进队
                if node.left is not None:
                    queue.append(node.left)
                # 右子树进队
                if node.right is not None:
                    queue.append(node.right)
            # 此层遍历完，进入下一层循环
            yield level_nodes


class TestSolution(unittest.TestCase):

    def setUp(self):
        self.test_case = [
            ([3, 9, 20, None, None, 15, 7], [
                [3],
                [9, 20],
                [15, 7]
            ]),
        ]
        self.s = Solution()

    def test_solution(self):
        for nums, answer in self.test_case:
            root = TreeNode.create(nums)
            ans = self.s.levelOrder(root)
            # self.assertEqual(answer,ans )


if __name__ == '__main__':
    unittest.main()

File: leetcode-in-rust/src/solution/_102_binary_tree_level_order_traversal.rs
use crate::datastruct::treenode::TreeNode;
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::VecDeque;
use std::option::Option::Some;

pub struct Solution {}

impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        if root.is_none() {
            return vec![];
        }
        let mut ans = vec![];
        let mut queue = VecDeque::new();
        queue.push_back(root);
        while !queue.is_empty() {
            let size = queue.len();
            let mut level = vec![];
            for _i in 0..size {
                if let Some(ref node) = queue.pop_front().flatten() {
                    let node = node.borrow();

                    level.push(node.val);

                    if let Some(ref left) = node.left {
                        queue.push_back(Some(Rc::clone(left)))
                    }
                    if let Some(ref right) = node.right {
                        queue.push_back(Some(Rc::clone(right)))
                    }
                }
            }
            ans.push(level);
        }
        ans
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solution() {
        let test_case = vec![
            (vec![Some(3), Some(9), Some(20), None, None, Some(15), Some(7)], vec![
                vec![3],
                vec![9, 20],
                vec![15, 7]
            ]),
        ];
        for (nums, answer) in test_case {
            let root = TreeNode::create(nums);
            let ans = Solution::level_order(root);
            assert_eq!(ans, answer);
        }
    }
}
